ppcs/ppc0022-metaprogramming.md

```diff
@@ -0,0 +1,385 @@
+
+The get-like functions come in pairs, consisting of a function named `get_...` and one called `can_...`. If the target item exists, then in either case it is returned. The behaviour differs when the item does not exist. The `get` variant will throw an exception, whereas the `can` variant will return `undef`. This allows the caller to decide which behaviour is best for the use-case.
+
+It is currently still an undecided question on what ought to be the correct behaviour for `add`-like functions when asked to create something that already exists, or the `remove`-like functions when asked to remove something that doesn't exist. It may be solved by something similar to above, where there are two differently-named functions that differ in that case. As yet there aren't any particularly compelling solutions.
+
+It is also undecided whether this API should offer an API that exposes the idea of globs as Perl implements them, or whether it should abstract those away to more end-user focused concepts. Some parts of the following API are labelled as optional on that basis.
+
+### Functions to operate on Packages
+
+These are functions to create, manipulate, and inspect packages themselves.
+
+#### `get_package`
+
+```perl
+$metapackage = meta::get_package($name);
+```
+
+Returns a meta-package object instance to represent the named package, if such a package exists. If not an exception is thrown.
+
+[[TODO: Decide whether "all packages exist", and thus this can never fail]]
+
+#### `add_package`
+
+```perl
+$metapackage = meta::add_package($name);
+```
+
+Creates a new package of the given name, and returns a meta-package object instance to represent it.
+
+#### `list_packages`
+
+```perl
+@metapackages = meta::list_packages($basename);
+```
+
+Returns a list of sub-packages within the given base package name, each as a meta-package object instance.
+
+### Functions to operate on References
+
+Some less common functions that may still be useful in some situations. These functions can be used to obtain meta-programming object instances in ways other than walking the symbol table for given names.
+
+#### `get_variable`
+
+```perl
+$metavar = meta::get_variable($varref);
+```
+
+Given a reference to a package variable, returns a meta-variable object instance to represent it. If `$varref` is not a SCALAR, ARRAY or HASH reference, an exception is thrown.
+
+*Note:* It is currently unspecified what happens if the reference refers to a lexical or anonymous variable, rather than a package one. This PPC does not specify any useful behaviour for such instances, but it's possible that future abilities may be added that become useful on lexical variables or anonymous containers.
+
+#### `get_subroutine`
```

Comment:
Yeah a fair question. I think originally I imagined it would want to be different because it returns a different class of meta- object to wrap the thing it refers to. But perhaps the same would be done for scalar/array/hash anyway, because maybe those metas have ways to operate on the structure of the underlying thing. Undecided.

But yes you're right, it doesn't have to be separate at all. Maybe best that it isn't - perhaps a single

```perl
$metavar = meta::for_reference( \$scalar | \@array | \%hash | \&code );
```

would suffice.

Reply:
made another attempt in 409275e